### DISCUSSION TITLE
Occasionally getting an error with black images

### DISCUSSION THREAD

**Discussion by brknsoul** [vladmandic/sdnext#1087] [Q&A] [answered]
I'm occasionally getting black images (or black faces using adetailer) with the following error in my prompt;

```
G:\automatic\modules\processing.py:641: RuntimeWarning: invalid value encountered in cast
  x_sample = x_sample.astype(np.uint8)
```

I imagine that might be due to a casting failure. These are my settings; (Win10/Chrome/AMD RX 6600 XT 8Gb/`--use-directml --medvram`)
![image](https://github.com/vladmandic/automatic/assets/1969381/156953f8-d717-45cc-b43d-027545da8da0)
![image](https://github.com/vladmandic/automatic/assets/1969381/efe3a5e5-a33d-4d45-bc63-1a64b6ebba7f)

**Comment by vladmandic** [comment]
its not a cast thing at all, its just that new `numpy` reports it slightly differently - its basically a `NaN` that occured inside image.
and since you have "do not check for NaN values" enabled, no check was performed before this code was reached.

you say this is happening occasionally - if this happening with a specific model only? can you reproduce at will or?

**Comment by brknsoul** [comment]
I was doing some XYZ plots yesterday to test samplers, unfortunately I deleted those images/grids, the ones I do have I used "aZovyaPhotoreal_v1Pruned" and "hardBLEND".

I couldn't reproduce it at will.

I'll fire up SD.Next and do another couple of plots to see if I get reproduce the error.

Ok, Still gives me errors;
Attached is the XYZ plot grid. I'm running a batch at Heun 25 to see if I can repro;
`Size: 512x512, Model hash: 0cbf567c8c, Model: aZovyaPhotoreal_v1Pruned, VAE: vae-ft-mse-840000-ema-pruned, Clip skip: 1, Version: 794b23c, Parser: Full parser`

![xyz_grid-0001-20230520222925-3480796489](https://github.com/vladmandic/automatic/assets/1969381/1220d9e0-c80d-46ca-8d8e-4c3834f558a1)

(PLMS _really_ sucks at low steps ;-) )

**Comment by brknsoul** [comment]
Heun, 25 steps
Batch 1, 10 images: 20% failure.
Batch 2, 10 images: 0% failure.
Batch 3, 10 images: 10% failure.

Seems to be somewhat random.

Hrmm.. iirc I never got errors using DPM++ SDE Karras.. lemme run a batch through that...

**Reply by vladmandic** [reply]
thats is REALLY RANDOM! :(
makes me wonder if there truly is a broken math somewhere, but nearly impossible to trace down if i cannot reproduce. which makes me wonder even more if broken math is actually inside directml itself since no such problems exist with cuda.

what's the version of `torch-directml` you have?

**Reply by brknsoul** [reply]
(torch-directml 0.2.0.dev230426)

ok, 30 images, 25 steps, DPM++ SDE Karras, 2 images failed.

Yeah, random af!

**Comment by vladmandic** [comment]
looping in @lshqqytiger as he's doing most of amazing work with directml - any ideas about this one?
unfortunately, it seems that directml team is not really using their github repo for anything other than posting releases as any open question/issues are not trigaged.

**Reply by brknsoul** [reply]
Interestingly that error in my first post only happens when plotting. The console is silent when generating a black image via normal generation.

**Reply by vladmandic** [reply]
since numpy raises warning, not an error, it is suppressed during normal procesing. xyz grid does not have that suppression.

**Comment by brknsoul** [comment]
Seems the error is model independent. Just tried a few runs on v1.5pruned, and got errors too.

**Comment by TeutonJon78** [comment, marked as answer]
In my testing on my RX580 with 8GB, subquadratic optimizations will often produce those black squares, especially in big batches/scripts.

Split attention is the option often recommended for AMD cards, (opt-split-attention-v1). It will give you no black squares, but severely limits the VRAM available for bigger resolutions or controlnets. I switched to doggettx (opt-split-attention) which gives zero black squares and also better VRAM usage for those other features. I tried SDP, but it fails the benchmark at batch size 2 and 4, while doggettx passes.

I also noticed in your settings you are using no-half and upcast sampling, but I think those are mutually exclusive options (vlad can correct me if I'm wrong). And i found with upcast sampling, again it wouldn't pass the benchmark test.

(I will say I did get one black square on doggettx in a batch of 512x768 for a 512x512 model, but it was also two images before the whole GPU black screen locked, so something was likely going wrong already.)

I also don't have upcast the layers to FP32 -- I don't know what that does.

**Reply by brknsoul** [reply]
I'll try those settings out.

I remember using `--opt-split-attention` (non v1) when using lshqq's fork with better results than --opt-split-attention-v1.

Perhaps @vladmandic could rename doggettx or put --opt-split-attention in parenthesis after it, for those moving from A1111 or lshqq's fork.

I noticed as well that I get more black images when batching, than with single generations.

Thanks for helping out with the testing.

**Comment by brknsoul** [comment]
Excellent! After using @TeutonJon78's settings, I got no errors/black images!

Settings used:
![image](https://github.com/vladmandic/automatic/assets/1969381/8159efff-c48b-47b2-bfb5-0f7dcaec1a17)
_(--no-half and --no-half-vae are being forced on, due to ~AMD's~ DirectML's inability to autocast)_
![image](https://github.com/vladmandic/automatic/assets/1969381/26232646-d10a-41a8-9e66-b02eff260842)


`a photo of vladmandic holding a cat`
Negatives: `a featureless colour shader, 3d max,`

![xyz_grid-0009-20230522123542-790195525](https://github.com/vladmandic/automatic/assets/1969381/e86cf315-9f5d-4db3-ab49-7215f02abce2)

_(I must say, you're looking quite handsome there, vlad! ;) )_

**Reply by vladmandic** [reply]
LMAO!!! :)

> AMD's inability to autocast

just to assign blame correctly - I blame AMD for not releasing ROCm for Windows, so you're forced to use DirectML instead - and DirectML cannot autocast :)

**Reply by TeutonJon78** [reply]
@vladmandic, can you comment about is upcast-sampling can also be on at the same time as no-half? There isn't any documentation I've found anywhere if they are complementary options or mutually exclusive.

And for us DirectML suffers, does the "upcast cross attention layer to FP32" do anything for us, required, or meaningless since we are forced into FP32 mode anyway?

**Reply by vladmandic** [reply]
- use fp16/bf16/fp32 means math will be done using that precision (works on device)
- no-half means that model will not be reduced to fp16 on load (works on model during load)
- no-half-vae means that vae will not be reduced to fp16 on load (works on model during load)
- upcast-sampling means that unet part of model will be upcast to fp32 (modifies model during execution)

all 4 can be used separately, but some combos make more sense than others
examples:
- if you have no-half, what's the point of fp16? not much except using more vram as model will be held in memory in fp32, but all math will be done on fp16.
- if you have no-half, what's the point of upcast from fp32 to fp32? none.
- what would be the point of using fp32 precision if model is reduced to fp16 during load? well, there are some differences in rounding precision (i can divide two fp16 numbers and get higher precision result in fp32, but you're not getting more precision from model itself, you're possibly eliminating gpu rounding errors).

now, all that would be simple **IF** all gpus correctly handled a) autocast (directml is does not have it), b) had full 16-bit precision when running fp16 (applies to most older gpus from both nvidia and amd, for example nvidia rtx1xxx/2xxx cards) instead of doing it wrong and resulting in ~14bit precision which is just not enough, so you need to run in full fp32 on some models which have weight values tuned too low.

h.t.h.

**Reply by brknsoul** [reply]
What's the difference between fp16 and bf16?

**Reply by lshqqytiger** [reply]
bfloat16 has more exponent bits than float16.
More: https://cloud.google.com/tpu/docs/bfloat16

**Reply by vladmandic** [reply]
different way of encoding floating point number into 16 bits - has higher min/max range, but less precision after decimal point.
for example, if you expect numbers between 0 and 1, you should use fp16 as you don't care about range and you get more precision. but if you expect extremely high values, bf16 is better. its a hit-or-miss as there is no way to know what numbers to expect.

![image](https://github.com/vladmandic/automatic/assets/57876960/db1c7e61-828b-46c0-a5f3-294636168062)

**Reply by TeutonJon78** [reply]
So how do "upcast cross attention layer to fp32" and "upcast sampling" interact?

**Reply by vladmandic** [reply]
ahh, forgot about that one.
`upcast_attn` runs same thing as `upcast-sampling` but in a different place: upscast sampling works on unet part of the model while upcast_attn works on cross-optimization (e.g., if you want xformers to run in fp32). i've never actually used that one, not sure what would be a good use case.

**Reply by TeutonJon78** [reply]
So, for us "forced to FP32" people we should really run them all, then? Just to make sure it's not trying to do any weird casting somewhere in the pipeline?

**Reply by Yodrik** [reply]
The note about being --half and --upcast being "either or" is from the automatic branch originally, but the note wasn't really well understood. A lot of us people started trying out the directml implementations without really understanding it. That was why I was curious how they affected things ([https://github.com/lshqqytiger/stable-diffusion-webui-directml/discussions/74#discussioncomment-5769940](url) 

I was OK myself up to mid May (I had occasional black images as well), but with latest commits, I seem to have lost my "sweet spot" and have nothing but OOM errors right now.

**Comment by ronbere** [comment]
![238102997-54e67be4-322f-4270-9f14-4529a3f1ee9e](https://github.com/vladmandic/automatic/assets/38155381/57cd5629-5e46-4b73-ac42-be9a35cc40e4)
this working for me
